# Libraries

In [1]:
import pandas as pd
import numpy as np

# 1.0. Exploratory Data Analysis

In [2]:
#Dataset upload from URL
url = 'https://raw.githubusercontent.com/lucasquemelli/House_Rocket/main/kc_house_data.csv'
data = pd.read_csv(url)

## 1.1. Data Description

In [3]:
#Visualization of the first 05 rows
data.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...      7        1180              0   
1      7242     2.0           0     0  ...      7        2170            400   
2     10000     1.0           0     0  ...      6         770              0   
3      5000     1.0           0     0  ...      7        1050            910   
4      8080     1.0           0     0  ...      8        1680              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   

   sqft_lot15  
0        5650  
1        7639  
2        8062  
3        5000  
4        7503  

[5 rows x 21 columns]

In [4]:
#Number of rows and columns of the dataset
print("Number of rows: {}\nNumber of columns: {}".format(data.shape[0],data.shape[1]))

Number of rows: 21613
Number of columns: 21


In [5]:
#Variable types and non-null counts
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             21613 non-null  int64  
 1   date           21613 non-null  object 
 2   price          21613 non-null  float64
 3   bedrooms       21613 non-null  int64  
 4   bathrooms      21613 non-null  float64
 5   sqft_living    21613 non-null  int64  
 6   sqft_lot       21613 non-null  int64  
 7   floors         21613 non-null  float64
 8   waterfront     21613 non-null  int64  
 9   view           21613 non-null  int64  
 10  condition      21613 non-null  int64  
 11  grade          21613 non-null  int64  
 12  sqft_above     21613 non-null  int64  
 13  sqft_basement  21613 non-null  int64  
 14  yr_built       21613 non-null  int64  
 15  yr_renovated   21613 non-null  int64  
 16  zipcode        21613 non-null  int64  
 17  lat            21613 non-null  float64
 18  long  

In [6]:
#Descriptive analysis
pd.set_option('display.float_format', lambda x: '%.2f' % x)
data.describe().T

count          mean           std        min           25%  \
id            21613.00 4580301520.86 2876565571.31 1000102.00 2123049194.00   
price         21613.00     540088.14     367127.20   75000.00     321950.00   
bedrooms      21613.00          3.37          0.93       0.00          3.00   
bathrooms     21613.00          2.11          0.77       0.00          1.75   
sqft_living   21613.00       2079.90        918.44     290.00       1427.00   
sqft_lot      21613.00      15106.97      41420.51     520.00       5040.00   
floors        21613.00          1.49          0.54       1.00          1.00   
waterfront    21613.00          0.01          0.09       0.00          0.00   
view          21613.00          0.23          0.77       0.00          0.00   
condition     21613.00          3.41          0.65       1.00          3.00   
grade         21613.00          7.66          1.18       1.00          7.00   
sqft_above    21613.00       1788.39        828.09     290.00       1190.00   
sqft_basement 21613.00        291.51        442.58       0.00          0.00   
yr_built      21613.00       1971.01         29.37    1900.00       1951.00   
yr_renovated  21613.00         84.40        401.68       0.00          0.00   
zipcode       21613.00      98077.94         53.51   98001.00      98033.00   
lat           21613.00         47.56          0.14      47.16         47.47   
long          21613.00       -122.21          0.14    -122.52       -122.33   
sqft_living15 21613.00       1986.55        685.39     399.00       1490.00   
sqft_lot15    21613.00      12768.46      27304.18     651.00       5100.00   

                        50%           75%           max  
id            3904930410.00 7308900445.00 9900000190.00  
price             450000.00     645000.00    7700000.00  
bedrooms               3.00          4.00         33.00  
bathrooms              2.25          2.50          8.00  
sqft_living         1910.00       2550.00      13540.00  
sqft_lot            7618.00      10688.00    1651359.00  
floors                 1.50          2.00          3.50  
waterfront             0.00          0.00          1.00  
view                   0.00          0.00          4.00  
condition              3.00          4.00          5.00  
grade                  7.00          8.00         13.00  
sqft_above          1560.00       2210.00       9410.00  
sqft_basement          0.00        560.00       4820.00  
yr_built            1975.00       1997.00       2015.00  
yr_renovated           0.00          0.00       2015.00  
zipcode            98065.00      98118.00      98199.00  
lat                   47.57         47.68         47.78  
long                -122.23       -122.12       -121.31  
sqft_living15       1840.00       2360.00       6210.00  
sqft_lot15          7620.00      10083.00     871200.00

Notice that there is no statistical meaning when applying descriptive analysis for "id", "zipcode", "lat" and "long". 

In [7]:
#Displaying unique values for the number of bedrooms, bathrooms, floors, waterfront, view, condition and grade
print("Unique values for bedrooms:", np.sort(data['bedrooms'].unique())) 
print("\nUnique values for bathrooms:", np.sort(data['bathrooms'].unique()))
print("\nUnique values for floors:", np.sort(data['floors'].unique()))
print("\nUnique values for waterfront:", np.sort(data['waterfront'].unique()))
print("\nUnique values for view:", np.sort(data['view'].unique()))
print("\nUnique values for condition:", np.sort(data['condition'].unique()))
print("\nUnique values for grade:", np.sort(data['grade'].unique()))

Unique values for bedrooms: [ 0  1  2  3  4  5  6  7  8  9 10 11 33]

Unique values for bathrooms: [0.   0.5  0.75 1.   1.25 1.5  1.75 2.   2.25 2.5  2.75 3.   3.25 3.5
 3.75 4.   4.25 4.5  4.75 5.   5.25 5.5  5.75 6.   6.25 6.5  6.75 7.5
 7.75 8.  ]

Unique values for floors: [1.  1.5 2.  2.5 3.  3.5]

Unique values for waterfront: [0 1]

Unique values for view: [0 1 2 3 4]

Unique values for condition: [1 2 3 4 5]

Unique values for grade: [ 1  3  4  5  6  7  8  9 10 11 12 13]


While there are properties with 33 bedrooms, also are there with 0. The same happens with the number of bathrooms.              This would properly be classified as inconsistences, yet I choose to classify into different types of properties in            this analysis. 

The other attributes contain a large number of unique values, such as "id" - as it may be seen below. Thus, they were not considered in this analysis. 

In [8]:
print("Number of unique values for 'id':", data['id'].nunique())

Number of unique values for 'id': 21436


While the number of rows is 21613, the number of unique "id" is 21436. It means there are only 21436 properties. 

## 1.2. Data Cleaning

In [13]:
#Checking for missing values
data.isna().sum()

id               0
date             0
price            0
bedrooms         0
bathrooms        0
sqft_living      0
sqft_lot         0
floors           0
waterfront       0
view             0
condition        0
grade            0
sqft_above       0
sqft_basement    0
yr_built         0
yr_renovated     0
zipcode          0
lat              0
long             0
sqft_living15    0
sqft_lot15       0
dtype: int64

In [10]:
#Checking for duplicates in identification number (id)
ids = data['id']
data[ids.isin(ids[ids.duplicated()])].sort_values('id')

id             date     price  bedrooms  bathrooms  \
2497      1000102  20150422T000000 300000.00         6       3.00   
2496      1000102  20140916T000000 280000.00         6       3.00   
16814     7200179  20141016T000000 150000.00         2       1.00   
16815     7200179  20150424T000000 175000.00         2       1.00   
11434   109200390  20141020T000000 250000.00         3       1.75   
...           ...              ...       ...       ...        ...   
6345   9828200460  20140627T000000 260000.00         2       1.00   
15200  9834200305  20150210T000000 615000.00         3       1.00   
15199  9834200305  20140716T000000 350000.00         3       1.00   
1085   9834200885  20140717T000000 360000.00         4       2.50   
1086   9834200885  20150420T000000 550000.00         4       2.50   

       sqft_living  sqft_lot  floors  waterfront  view  ...  grade  \
2497          2400      9373    2.00           0     0  ...      7   
2496          2400      9373    2.00           0     0  ...      7   
16814          840     12750    1.00           0     0  ...      6   
16815          840     12750    1.00           0     0  ...      6   
11434         1480      3900    1.00           0     0  ...      7   
...            ...       ...     ...         ...   ...  ...    ...   
6345           700      4800    1.00           0     0  ...      7   
15200         1790      3876    1.50           0     0  ...      7   
15199         1790      3876    1.50           0     0  ...      7   
1085          2080      4080    1.00           0     0  ...      7   
1086          2080      4080    1.00           0     0  ...      7   

       sqft_above  sqft_basement  yr_built  yr_renovated  zipcode   lat  \
2497         2400              0      1991             0    98002 47.33   
2496         2400              0      1991             0    98002 47.33   
16814         840              0      1925             0    98055 47.48   
16815         840              0      1925             0    98055 47.48   
11434        1480              0      1980             0    98023 47.30   
...           ...            ...       ...           ...      ...   ...   
6345          700              0      1922             0    98122 47.61   
15200        1090            700      1904             0    98144 47.58   
15199        1090            700      1904             0    98144 47.58   
1085         1040           1040      1962             0    98144 47.57   
1086         1040           1040      1962             0    98144 47.57   

         long  sqft_living15  sqft_lot15  
2497  -122.21           2060        7316  
2496  -122.21           2060        7316  
16814 -122.21           1480        6969  
16815 -122.21           1480        6969  
11434 -122.37           1830        6956  
...       ...            ...         ...  
6345  -122.30           1440        4800  
15200 -122.29           1360        4080  
15199 -122.29           1360        4080  
1085  -122.29           1340        4080  
1086  -122.29           1340        4080  

[353 rows x 21 columns]

Duplicates (353) in this dataset are meant to be properties which were sold twice or more between 2014 and 2015. These properties were sold on different dates and at different prices. It means the price changes with time. 

It is important to mention that the number of duplicates is different from the number of ids that are duplicates. Notice that the same id appears twice or more. Thus, the number of ids that are duplicate is: total number of rows - number of unique ids = 177.